lib/ViewModels/MapNavigation/ToolButton.ts

```diff
@@ -0,0 +1,152 @@
+import { action, computed } from "mobx";
+import createGuid from "terriajs-cesium/Source/Core/createGuid";
+import TerriaError from "../../Core/TerriaError";
+import ViewerMode from "../../Models/ViewerMode";
+import ViewState from "../../ReactViewModels/ViewState";
+import MapNavigationItemController from "./MapNavigationItemController";
+import { BaseButtonOptions } from "./MapToolbar";
+
+export interface ToolButtonOptions<ToolProps = any> extends BaseButtonOptions {
+  tool: {
+    name: string;
+    component: () =>
+      | React.ComponentType<ToolProps>
+      | Promise<React.ComponentType<ToolProps>>;
+    props?: ToolProps;
+  };
+}
+
+/**
+ * ToolButton
+ *
+ */
+export interface ToolButton<ToolProps = any> {
+  /**
+   * Removes the button from the toolbar
+   */
+  removeButton: () => void;
+
+  /**
+   * Closes the tool if it is open
+   */
+  closeTool: () => void;
+
+  /**
+   * Can be used to programatically open the tool
+   *
+   * @param props Additional Props to pass to the tool. This is useful, for
+   * example, if you want to open the tool in a specific mode.
+   */
+  openTool: (props?: Partial<ToolProps>) => void;
+}
+
+/**
+ * Add a button to open a Tool UI to the map navigation menu.
+ *
+ * This is useful for external plugins to add buttons to the navigation menu that launches some custom work flow.
+ *
+ * @param viewState The ViewState instance
+ * @param buttonOptionsGenerator A function that when called, returns {@link ToolButtonOptions} for the new button
+ * @returns A {@link ToolButton} instance
+ */
+export function addToolButton<ToolProps = any>(
+  viewState: ViewState,
+  buttonOptionsGenerator: (
+    button: ToolButton<ToolProps>
+  ) => ToolButtonOptions<ToolProps>
+): ToolButton<ToolProps> {
+  let controller: ToolButtonController<ToolProps> | undefined = undefined;
+  const toolButton = {
+    removeButton: () => terria.mapNavigationModel.remove(id),
+    closeTool: () => controller?.closeTool(),
+    openTool: (props?: Partial<ToolProps>) => {
+      const navigationItem = terria.mapNavigationModel.findItem(id);
+      const controller = navigationItem?.controller;
+      if (controller instanceof ToolButtonController) {
+        controller.openTool(props);
+      }
+    }
```

Comment:
The only reason I can think of now why I wrote it that way is a bit of defensive coding. If something else has removed the item from `mapNavigationModel` you still want `closeTool()` to work, so I use the closure value which will always exist. But, you want `openTool()` to fail if something else removed the button from `mapNavigationModel`, so I use the queried result. 

If it makes the code a bit simpler, I'm happy to make it symmetric and do a query in both cases (and maybe extract the query into a helper function)

Reply:
I have made the changes here - https://github.com/TerriaJS/terriajs/pull/6646/commits/72090a02e39d93e398ce7bbf09db1ba763db441d

Reply:
> The only reason I can think of now why I wrote it that way is a bit of defensive coding. If something else has removed the item from `mapNavigationModel` you still want `closeTool()` to work, so I use the closure value which will always exist. But, you want `openTool()` to fail if something else removed the button from `mapNavigationModel`, so I use the queried result.

Defensive coding is good, I think in this case a one line comment along the lines of:
```
// Access controller through mapNavigationModel to ensure the button hasn't been removed 
```

I think a comment similar to this would still be useful in `getController`.